mindmeld/components/entity_resolver.py

```diff
@@ -51,51 +65,189 @@ class EntityResolver:
     """
 
-    # prefix for Elasticsearch indices used to store synonyms for entity resolution
-    ES_SYNONYM_INDEX_PREFIX = "synonym"
-    """The prefix of the ES index."""
-
-    def __init__(
-        self, app_path, resource_loader, entity_type, es_host=None, es_client=None
-    ):
-        """Initializes an entity resolver
+    @classmethod
+    def validate_resolver_name(cls, name):
+        if name not in ENTITY_RESOLVER_MODEL_TYPES:
+            msg = "Expected 'model_type' in ENTITY_RESOLVER_CONFIG among {!r}"
+            raise Exception(msg.format(ENTITY_RESOLVER_MODEL_TYPES))
+        if name == "sbert_cosine_similarity" and not sbert_available:
+            raise ImportError(
+                "Must install the extra [bert] to use the built in embbedder for entity "
+                "resolution. See https://www.mindmeld.com/docs/userguide/getting_started.html")
+
+    def __new__(cls, app_path, resource_loader, entity_type, **kwargs):
+        """Identifies appropriate entity resolver based on input config and
+        initializes it.
 
         Args:
-            app_path (str): The application path
-            resource_loader (ResourceLoader): An object which can load resources for the resolver
-            entity_type: The entity type associated with this entity resolver
-            es_host (str): The Elasticsearch host server
+            app_path (str): The application path.
+            resource_loader (ResourceLoader): An object which can load resources for the resolver.
+            entity_type (str): The entity type associated with this entity resolver.
+            er_config (dict): A classifier config
+            es_host (str): The Elasticsearch host server.
+            es_client (Elasticsearch): The Elasticsearch client.
         """
-        self._app_namespace = get_app_namespace(app_path)
-        self._resource_loader = resource_loader
-        self._normalizer = resource_loader.query_factory.normalize
+        er_config = (
+            kwargs.pop("er_config", None) or
+            get_classifier_config("entity_resolution", app_path=app_path)
+        )
+        name = er_config.get("model_type", None)
+        cls.validate_resolver_name(name)
+        return ENTITY_RESOLVER_MODEL_MAPPINGS.get(name)(
+            app_path, resource_loader, entity_type, er_config, **kwargs
+        )
+
+    def fit(self):
+        raise NotImplementedError
+
+    def load(self):
+        raise NotImplementedError
+
+    def predict(self, entity):
+        raise NotImplementedError
+
+
+class EntityResolverBase(ABC):
+    """
+    Base class for Entity Resolvers
+    """
+
+    def __init__(self, app_path, resource_loader, entity_type, er_config, **kwargs):
+        """Initializes an entity resolver"""
+        self.app_path = app_path
+        self.resource_loader = resource_loader
         self.type = entity_type
+        self.er_config = er_config
+        self.kwargs = kwargs
```

Comment:
Why cant we enumerate all the fields from kwargs that we want to use later in this instance's functions?

Reply:
I kept it like that because different derived classes might require different field. For eg. Elastic Search requires `es_host` and `es_client` which will not be required by BERT based class or Exact match class.